3/models/autoencoder.py

```diff
@@ -0,0 +1,107 @@
+import torch.nn as nn
+import torch.nn.functional as F
+import torch
+import os
+import numpy as np
+from tensorboardX import SummaryWriter
+
+
+
+class AE(nn.Module):
+    """
+    Denoising autoencoder
+    """
+    def __init__(self, in_dim=39, botleneck_dim=20):
+        super(AE, self).__init__()
+        self.in_dim = in_dim
+        self.encoder = nn.Sequential(
+            nn.Linear(in_dim, 32),
+            nn.ReLU(True),
+            nn.Linear(32, 20),
+            nn.ReLU(True),
+            nn.Linear(20, botleneck_dim))
+        self.decoder = nn.Sequential(
+            nn.Linear(botleneck_dim, 20),
+            nn.ReLU(True),
+            nn.Linear(20, 32),
+            nn.ReLU(True),
+            nn.Linear(32, in_dim))
+
+    def forward(self, x):
+        x = self.encoder(x)
+        x = self.decoder(x)
+        return x
+
+
+class AETranner:
+
+    """
+    Trainner for AE
+    """
+
+    def __init__(self, model, device='cpu', model_dir='ckpt', log_dir='logs'):
+        self.device = device
+        self.model_dir = model_dir
+        self.model = model
+        self.model.to(self.device)
+        self.writer = SummaryWriter(log_dir=log_dir)
+
+    def save(self, epoch, train_name):
+        os.makedirs(self.model_dir, exist_ok=True)
+        torch.save(self.model.state_dict(), os.path.join(self.model_dir, f'model{train_name}_{epoch}.pt'))
+
+    def train(self, train_loader, lr=1e-4, n_epoch=25, train_name=None,  loss_point_every=5):
+        """
+        train AE
+        :param train_loader: train data (noise, clean)
```

Comment:
Looks nice
You you like giving comprehensive annotations for your functions explore typing in python in spare time